### PR TITLE
Implement migrations; fix #153

### LINKED ISSUE #153: Create SQL schema migration method
Migrating the sql schema will suck if it's not managed.

New table: `cockatrice_schema`

version int  | comment str
--- | --- 
1 | Current `servatrice.sql`

Add a new directory in git (`servatrice/migrations`) with servatrice.sql -> `1.sql`.

New migrations get added by:
`2.patch.sql` - This file contains `ALTER TABLE` statements to translate from version 1 to version 2
`2.create.sql` - This file contains `CREATE TABLE` statements to build this version of the db from scratch in its entirety to this version.

New script to upgrade to the latest schema.
- Read current version from db
- scan for new patches
- lock whole db
- snapshot db minus volatile tables
- start transaction
- apply each sql file in order
- create a scratch db and apply the full create sql
- compare updated schema against the scratch db
- if schema doesn't match, error out and abort transaction
- if it matches, drop the scratch schema
- unlock

Add servatrice install script that uses the highest-version create schema sql.

cc @woogerboy21

### PR DESCRIPTION
**Pull request by ctrlaltca**
Well, #153 also included implementing a script that automatically detect the version, snapshots and update the database schema.
I'm unsure wether making this automatic is a good idea, and we should probably make it support all the three major OSes..

### CODE REVIEW

**Comment by ctrlaltca** [comment]
A note about file names proposed in #153:
 * ```2.patch.sql``` is nice, but with more than 9 migrations, they'll start being wrongly ordered by ls, file managers, etc.., thus the padding.
 * ```2.create.sql``` means we should bundle all the old versions of the base schema file? imho it's useless at best, since it could lead users to apply the wrong version of base schema.

**Comment by woogerboy21** [comment]
is it a good idea to have servatrice create the schema table or would it be more appropriate to simply fix the sql file to have the basic "it table not exists" for all the tables?  I could see a scenario were some one hasnt updated the database properly yet servatrice builds the schema table and its not the correct db schema

**Comment by ctrlaltca** [comment]
I'm against having servatrice apply schema changes automatically: it's just too dangerous.

**Comment by woogerboy21** [comment]
@ctrlaltca agree'd.  I miss-read the files changed and thought that you added the ability for servatrice to create the schema table if it wasn't present.  doing so would be a bad thing :+1:

**Comment by Daenyth** [comment]
Rather than bundling all versions why don't we just keep the migration
scripts plus a master from scratch script.

What about downgrades?

**Comment by woogerboy21** [comment]
@Daenyth Downgrades?  Wouldnt that be the same as using a previous version of the servatrice application and only applying the necessary .sql files to the level of db needed?  Or were you thinking there would be some type of need for some .sql script that undoes what that level of schema version .sql script does?

**Comment by Daenyth** [comment]
A script. consider if someone upgrades their server, realizes there is a
bug, and wants to downgrade - the schema could potentially be incompatible.
Deleting the database and rebuilding is undesirable because you'd lose the
data in it.

**Comment by woogerboy21** [comment]
Or the user could dump the data out and back into the new database they created using the "master from scratch" script on the particular version of schema they desire and then change the db name in the ini and restart the server.  I don't think I would ever recommend "downgrading" a db without having a backup.  And if I have a backup I would recommend creating a new DB and importing the backup information over using a downgrade.

But that is just my opinion I guess.....

**Comment by ctrlaltca** [comment]
Downgrades are hard.. Imagine having to add back a column after dropping it, or having to add back an unique key when your table data is not unique anymore, like #1119: you just can't do it.
The best solution is making a backup before any schema change and keep them handy.

**Comment by Daenyth** [comment]
Fair enough

**Comment by woogerboy21** [comment]
Does the version check done allow for point releases to a db scheme rather than major number only? I can see a use for having scheme version numbers like 2.1 if all your doing is adding a field to a table or something minor.

File: servatrice/src/servatrice_database_interface.cpp [hunk 1]
```diff
@@ -60,6 +60,25 @@ bool Servatrice_DatabaseInterface::openDatabase()
...
+        qCritical() << QString("[%1] Error opening database: unable to load database schema version (hint: ensure the cockatrice_schema_version exists)").arg(poolStr);
+        return false;
+    }
+
+    if (versionQuery->next()) {
+        const int dbversion = versionQuery->value(0).toInt();
+        const int expectedversion = DATABASE_SCHEMA_VERSION;
+        if(dbversion != expectedversion)
+        {
+            qCritical() << QString("[%1] Error opening database: the database schema version is too old, yum need to run the migrations to update it from version %2 to version %3").arg(poolStr).arg(dbversion).arg(expectedversion);
```
Comment (woogerboy21) [inline servatrice/src/servatrice_database_interface.cpp]: yum need should read you need

**Comment by woogerboy21** [comment]
@ctrlaltca How about this one?  I think this one looks complete.  I have not had a chance to test it but i dont believe there is any additional functionality you would like to add is there?

**Comment by ctrlaltca** [comment]
It's complete to me

**Comment by ZeldaZach** [comment]
I'm not really certain what this PR does. Can you explain quickly then I'll merge?

**Comment by ctrlaltca** [comment]
Servatrice can use a database to store players, rooms, etc.
The database schema can change when we update servatrice with new features (eg. adding a new property for users).
When an admin updates servatrice, he needs to adapt his existing database to the new version.
In order to do this, he'll need a migration script.
By now we only provided a "create a new schema script", but this is not really admin-friendly, because it forces admins to manually search for schema changes and apply them.

This PR adds a base folder to store migration scripts, and adds a "database version" check within servatrice so that it will refuse to start and notify the admin if he needs to update the database schema:
```
"[main] Opening database..."
"[main] Error opening database: the database schema version is too old, yum need to run the migrations to update it from version 0 to version 1"
Failed to open database
```

**Comment by ZeldaZach** [comment]
Seems like a good idea to me.

**Pull request merged**